### DISCUSSION TITLE
couple of bash functions might be useful

### DISCUSSION THREAD

**Discussion by rleaver152** [policratus/pupyl#129] [Show and tell]
These might save some time for people - I put these in my .bash_aliases file (Mac OSX).   They assume a virtual environment in the pupyl-venv directory  as set out.

The first builds the data directory named after the source top directory
The second serves from that same source.

In both cases they need the source directory as an argument.   Hope this helps people.   

```
python3.9 -m venv pupyl-venv
```

```
reverse_search_images_my_pupyl_build() {
target_directory="$(basename "$1")"
source $HOME/Scripts/PythonEnvironments/pupyl-venv/bin/activate
pupyl_directory="$HOME/.pupyl""_""$target_directory"
reference_images_directory="$1"
pupyl --data_dir "$pupyl_directory" index "$reference_images_directory"
deactivate
}
export -f reverse_search_images_my_pupyl_build

reverse_search_images_my_pupyl() {
target_directory="$(basename "$1")"
source $HOME/Scripts/PythonEnvironments/pupyl-venv/bin/activate
pupyl_directory="$HOME/.pupyl""_""$target_directory"
reference_images_directory="$1"
pupyl --data_dir "$pupyl_directory" serve
deactivate
}
export -f reverse_search_images_my_pupyl


```

**Comment by policratus** [comment]
Hey @rleaver152, thanks for the discussion!

If I really understood your script, the first one (`reverse_search_images_my_pupyl_build`) is something like:

```shell
pupyl --data_dir /path/to/your/data/dir index /path/to/images/
```

and the second one (`reverse_search_images_my_pupyl`):

```shell
pupyl --data_dir /path/to/your/data/dir serve
```

, is that right?

**Comment by rleaver152** [comment]
exactly - with the advantage of persisting named inditdual data databases

**Comment by policratus** [comment]
Hey @rleaver152, hence the following can achieve exactly what your script is doing:

```shell
pupyl --data_dir /path/to/your/data/dir index /path/to/images/
pupyl --data_dir /path/to/your/data/dir serve
```

Could you test it and return with your feedback? Thanks a lot!

**Reply by rleaver152** [reply]
hi - but its already in my scripts!
....bash
 pupyl --data_dir "$pupyl_directory" index "$reference_images_directory"
..... bash
The reason I call from bash is that I use virtual environments (my MacOSX base python environment is untouched/sparse to prevent issues with multiple python variations).   This way I can use standard CLI to invoke the program

**Reply by policratus** [reply]
Oh, sorry @rleaver152, you're right! It's a great way to just execute `pupyl` CLI, without the need to activate a virtual environment. Great work!

**Comment by rleaver152** [comment]
the real *wish list* is outputting the top [10] near matches as discussed a few weeks back.  That would be awesome!      Is it possible?  (wish I knew how to do it!)

**Comment by policratus** [comment]
Yep, for sure @rleaver152 this is possible! We just released `pupyl` version [`0.13.0`](https://github.com/policratus/pupyl/releases/tag/v0.13.0) and this is the next thing we'll be working on.

**Comment by policratus** [comment]
@rleaver152, this thread will be addressed also in #123.

**Reply by policratus** [reply]
Hi @rleaver152, just opened a pull request (#133) implementing this idea.
